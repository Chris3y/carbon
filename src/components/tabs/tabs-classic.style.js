import { css } from 'styled-components';
import { THEMES } from '../../style/themes';

export default ({ theme }) => theme.name === THEMES.classic
  && css`
    ${({ position }) => position === 'top'
      && css`
        margin-top: 15px;
      `}

    ${({ position }) => position === 'left'
      && css`
        display: flex;
        width: 100%;
      `}
  `;
