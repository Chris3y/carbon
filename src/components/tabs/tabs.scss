@import "./../../style-config/colors";
@import "./../../style-config/general";

$carbon-tabs__header__background-color: #f5f6f7 !default;
$carbon-tabs__header__border-color: $grey-dark-blue-20 !default;
$carbon-tabs__header--selected_border-color: $blue-bright !default;
$carbon-tabs__header__color: $grey-dark !default;

.carbon-tabs__headers__header--error {
  border-bottom-color: $red;
}

.carbon-tabs__headers__header--warning {
  border-bottom-color: $orange-mid;
}
