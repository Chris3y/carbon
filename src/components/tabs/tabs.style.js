import styled, { css } from 'styled-components';
import tabsClassicStyle from './tabs-classic.style';

const StyledTabs = styled.div`
${({ position }) => position === 'top'
  && css`
    margin-top: 15px;
  `}

${({ position }) => position === 'left'
  && css`
    display: flex;
    width: 100%;
  `}
  ${tabsClassicStyle}
`;

export default StyledTabs;
