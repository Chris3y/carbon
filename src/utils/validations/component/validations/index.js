import formWithValidation from './form-with-validations.hoc';
import withUniqueName from './with-unique-name.hoc';
import withValidation from './with-validation.hoc';

export { formWithValidation, withUniqueName, withValidation };
export default formWithValidation;
