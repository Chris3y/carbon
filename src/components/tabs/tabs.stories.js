import React from 'react';
import { storiesOf } from '@storybook/react';
import { select } from '@storybook/addon-knobs';
import { Tab, Tabs } from './tabs';
import OptionsHelper from '../../utils/helpers/options-helper';
import notes from './notes.md';
import { StoryHeader, StoryCode, StoryCodeBlock } from '../../../.storybook/style/storybook-info.styles';

storiesOf('Tabs', module)
  .addParameters({
    info: {
      propTablesExclude: []
    }
  })
  .add(
    'default',
    () => {
      const selectOption = ['top', 'left'];
      const align = select('align', OptionsHelper.alignBinary, Tabs.defaultProps.align);
      const position = select('position', selectOption, Tabs.defaultProps.position);

      return (
        <Tabs align={ align } position={ position }>
          <Tab tabId='tab-1' title='Tab 1'>
          Content for tab 1
          </Tab>
          <Tab tabId='tab-2' title='Tab 2'>
          Content for tab 2
          </Tab>
        </Tabs>
      );
    },
    {
      info: {
        text: (
          <div>
            <p>Tabs component</p>

            <StoryHeader> Implementation</StoryHeader>

            <p>Import the component:</p>
            <StoryCode padded>{'import { Tabs, Tab } from "carbon-react/lib/components/tabs"'}</StoryCode>

            <p>To render Tabs component: </p>
            <StoryCodeBlock>
              {'<Tabs>'}
              {'  <Tab title="Title 1" tabId="uniqueId1">'}

              {'    <Textbox />'}
              {'    <Textbox />'}

              {'  </Tab>'}

              {'  <Tab title="Title 2" tabId="uniqueId2">'}

              {'    <Date />'}
              {'  <Textbox />'}

              {'  </Tab>'}
              {'</Tabs>'}
            </StoryCodeBlock>

            <p>
            Optionally, you can pass
              <StoryCode padded>renderHiddenTabs</StoryCode>
            prop to the Tabs.
            </p>
            <p>
            By default this is set to
              <StoryCode padded>true</StoryCode>
            and therefore all tabs will be rendered.
            </p>
            <p>
            The selected tab will have a class of
              <StoryCode padded>selected</StoryCode>
            and all other tabs will have a class of
              <StoryCode padded>hidden</StoryCode>
            which sets their display to
              <StoryCode padded>none</StoryCode>.
            </p>

            <p>
            Setting
              <StoryCode padded>renderHiddenTabs</StoryCode> to false will add a small performance improvement as all
            previously hidden tabs will not be rendered to the page.
            </p>
            <p>
            If you are using the tab component within a form all tabs should be rendered so that form validation can
            work correctly.
            </p>
            <p>
            The tabs widget also allows you to select a tab on page load. By default this is set to the first tab. To
            set a different tab on page load pass a tabId to the selectedTabId prop as shown in the example below.
            </p>
            <p>
          To render a Tabs Widget with Options:
            </p>
            <StoryCodeBlock>
              {'<Tabs renderHiddenTabs={ false } selectedTabId="uniqueId2" >'}
              {'  <Tab title="Title 1" tabId="uniqueId1">'}

              {'    <Textbox />'}
              {'    <Textbox />'}

              {'  </Tab>'}

              {'  <Tab title="Title 2" tabId="uniqueId2">'}

              {'    <Date />'}
              {'    <Textbox />'}

              {'  </Tab>'}
              {'</Tabs>'}
            </StoryCodeBlock>
          </div>
        )
      }
    },
    { notes: { markdown: notes } }
  );
