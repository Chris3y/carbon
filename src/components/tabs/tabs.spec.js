import React from 'react';
import TestUtils from 'react-dom/test-utils';
import Immutable from 'immutable';
import { shallow, mount } from 'enzyme';
import 'jest-styled-components';
import { Tabs, Tab } from './tabs.component';
import Textbox from '../textbox';
import { elementsTagTest, rootTagTest } from '../../utils/helpers/tags/tags-specs';
import Browser from '../../utils/helpers/browser';
import TabsHeader from './tabs-header/tabs-header.component';

function render(props) {
  return shallow(
    <Tabs { ...props }>
      <Tab title='Tab Title 1' tabId='uniqueid1'>
        TabContent
      </Tab>
      <Tab title='Tab Title 2' tabId='uniqueid2'>
        TabContent
      </Tab>
      <Tab title='Tab Title 3' tabId='uniqueid3'>
        TabContent
      </Tab>
    </Tabs>
  );
}

function renderStyles(props) {
  return TestRenderer.create(<StyledTabHeader
    title='Tab Title 1' id='uniqueid1'
    { ...props }
  />);
}

describe('Tabs', () => {
  let instance;
  const instanceWithNull = TestUtils.renderIntoDocument(
    <Tabs renderHiddenTabs={ false }>
      {null}
      <Tab title='Tab Title 1' tabId='uniqueid1'>
        <Textbox name='foo' />
        <Textbox name='bar' />
      </Tab>
    </Tabs>
  );

  const instanceOneChild = TestUtils.renderIntoDocument(
    <Tabs renderHiddenTabs={ false }>
      <Tab title='Tab Title 1' tabId='uniqueid1'>
        <Textbox name='foo' />
        <Textbox name='bar' />
      </Tab>
    </Tabs>
  );

  beforeEach(() => {
    instance = TestUtils.renderIntoDocument(
      <Tabs>
        <Tab
          title='Tab Title 1' tabId='uniqueid1'
          className='class1' headerClassName='headerClass1'
        >
          <Textbox name='foo' />
          <Textbox name='bar' />
        </Tab>
        <Tab
          title='Tab Title 2' tabId='uniqueid2'
          className='class2' headerClassName='headerClass2'
        >
          <Textbox name='baz' />
          <Textbox name='bax' />
        </Tab>
        <Tab
          title='Tab Title 3'
          tabId='uniqueid3'
          className='class3 class4'
          headerClassName='headerClass3 headerClass4'
        >
          <Textbox name='bar' />
          <Textbox name='bap' />
        </Tab>
      </Tabs>
    );
    instance._window = {
      history: {
        replaceState: () => {}
      },
      location: ''
    };
  });

  describe('initialize', () => {
    it('sets tabValidity to a empty immutable map', () => {
      expect(instance.state.tabValidity).toEqual(Immutable.Map());
    });
    it('sets tabWarning to a empty immutable map', () => {
      expect(instance.state.tabWarning).toEqual(Immutable.Map());
    });
  });

  describe('componentWillMount', () => {
    describe('initial selected tab', () => {
      describe('when passed as props', () => {
        it('uses the prop as selectedTabId and takes precendent over hash', () => {
          instance = TestUtils.renderIntoDocument(
            <Tabs selectedTabId='uniqueid2'>
              <Tab title='Tab Title 1' tabId='uniqueid1'>
                <Textbox name='foo' />
                <Textbox name='bar' />
              </Tab>
              <Tab title='Tab Title 2' tabId='uniqueid2'>
                <Textbox name='baz' />
                <Textbox name='bax' />
              </Tab>
            </Tabs>
          );
          instance._window = { location: '#uniqueid1' };

          expect(instance.state.selectedTabId).toEqual('uniqueid2');
        });
      });

      describe('detects a hash', () => {
        beforeEach(() => {
          instance = TestUtils.renderIntoDocument(
            <Tabs>
              <Tab title='Tab Title 1' tabId='uniqueid1'>
                <Textbox name='foo' />
                <Textbox name='bar' />
              </Tab>
              <Tab title='Tab Title 2' tabId='uniqueid2'>
                <Textbox name='baz' />
                <Textbox name='bax' />
              </Tab>
            </Tabs>
          );
        });

        it('matches the hash so uses the tab', () => {
          instance._window = { location: { hash: '#uniqueid2' } };
          instance.componentWillMount();
          expect(instance.state.selectedTabId).toEqual('uniqueid2');
        });

        it('does not match the hash so uses first tab', () => {
          instance._window = { location: { hash: '#foo' } };
          instance.componentWillMount();
          expect(instance.state.selectedTabId).toEqual('uniqueid1');
        });
      });

      describe('when not passed as props', () => {
        it('it defaults to the first child', () => {
          expect(instance.state.selectedTabId).toEqual('uniqueid1');
        });
      });

      describe('when there is only one tab', () => {
        it('uses the first tab', () => {
          expect(instanceOneChild.state.selectedTabId).toEqual('uniqueid1');
        });
      });

      describe('when passed a null child', () => {
        it('ignores the null child', () => {
          expect(instanceWithNull.state.selectedTabId).toEqual('uniqueid1');
        });
      });
    });
  });

  describe('Change in tab prop', () => {
    let instance, tabs, unique1Tab, unique2Tab;
    beforeEach(() => {
      spyOn(Browser, 'getWindow').and.returnValue({
        history: {
          replaceState: () => {}
        },
        location: ''
      });
      class DummyComponent extends React.Component {
        state = {
          selectedTabId: 'uniqueid2',
          align: 'left'
        };

        render() {
          return (
            <Tabs { ...this.state }>
              <Tab title='Tab Title 1' tabId='uniqueid1'>
                <Textbox name='foo' />
                <Textbox name='bar' />
              </Tab>
              <Tab title='Tab Title 2' tabId='uniqueid2'>
                <Textbox name='baz' />
                <Textbox name='bax' />
              </Tab>
            </Tabs>
          );
        }
      }

      const TestParent = React.createFactory(DummyComponent);
      instance = TestUtils.renderIntoDocument(TestParent());
      tabs = TestUtils.scryRenderedComponentsWithType(instance, Tab);
      unique1Tab = tabs[0];
      unique2Tab = tabs[1];
      expect(unique1Tab.props.isTabSelected).toEqual(false);
      expect(unique2Tab.props.isTabSelected).toEqual(true);
    });

    describe('without noTabChange function', () => {
      it('changes tab to current prop', () => {
        instance.setState({
          selectedTabId: 'uniqueid1'
        });
        expect(unique2Tab.props.isTabSelected).toEqual(false);
        expect(unique1Tab.props.isTabSelected).toEqual(true);
      });

      it('change in other tab', () => {
        instance.setState({
          align: 'right'
        });
        expect(unique1Tab.props.isTabSelected).toEqual(false);
        expect(unique2Tab.props.isTabSelected).not.toEqual(false);
      });
    });

    describe('with onTabChange function', () => {
      let onClick;
      beforeEach(() => {
        onClick = jasmine.createSpy('tab change');
        instance.setState({
          onTabChange: onClick
        });
      });

      it('calls onTabChange function', () => {
        instance.setState({
          selectedTabId: 'uniqueid1'
        });
        expect(onClick).toHaveBeenCalledWith('uniqueid1');
      });
    });

    describe('when tab already clicked to new prop', () => {
      it('does not change tab', () => {
        const tabs = TestUtils.findRenderedComponentWithType(instance, Tabs);
        spyOn(tabs, 'updateVisibleTab').and.callThrough();
        const tabHeaders = TestUtils.scryRenderedDOMComponentsWithClass(instance, 'carbon-tabs__headers__header');
        TestUtils.Simulate.click(tabHeaders[0], { target: { dataset: { tabid: 'uniqueid1' } } });
        expect(tabs.updateVisibleTab).toHaveBeenCalled();

        tabs.updateVisibleTab.calls.reset();

        instance.setState({
          selectedTabId: 'uniqueid1'
        });
        expect(tabs.updateVisibleTab).not.toHaveBeenCalled();
      });
    });
  });

  describe('changeValidity', () => {
    beforeEach(() => {
      instance.setState({ tabValidity: Immutable.fromJS({ foo: true }) });
      spyOn(instance, 'setState').and.callThrough();
      instance.changeValidity('foo', false);
    });

    it('sets the validity state for the given tab', () => {
      expect(instance.state.tabValidity.toJS()).toEqual({ foo: false });
    });
  });

  describe('changeWarning', () => {
    beforeEach(() => {
      instance.setState({ tabWarning: Immutable.fromJS({ foo: false }) });
      spyOn(instance, 'setState').and.callThrough();
      instance.changeWarning('foo', true);
    });

    it('sets the warning state for the given tab', () => {
      expect(instance.state.tabWarning.toJS()).toEqual({ foo: true });
    });
  });

  describe('handleTabClick', () => {
    it('sets the state to the currently selected tabId', () => {
      spyOn(instance, 'handleTabClick');
      const secondTab = TestUtils.scryRenderedDOMComponentsWithTag(instance, 'li')[1];

      expect(secondTab.classList.contains('carbon-tabs__headers__header--selected')).toBeFalsy();

      TestUtils.Simulate.click(secondTab, { target: { dataset: { tabid: 'uniqueid2' } } });

      expect(secondTab.classList.contains('carbon-tabs__headers__header--selected')).toBeTruthy();
    });

    it('sets the location', () => {
      const replaceSpy = jasmine.createSpy('replaceState');
      instance._window = {
        history: {
          replaceState: replaceSpy
        },
        location: {
          origin: 'foo',
          pathname: 'bar'
        }
      };
      instance.handleTabClick({ target: { dataset: { tabid: 'foo' } } });
      expect(replaceSpy).toHaveBeenCalledWith(null, 'change-tab', 'foobar#foo');
    });

    describe('when a onTabChange prop is passed', () => {
      it('calls the prop', () => {
        const clickSpy = jasmine.createSpy('tabClick');

        const instance = TestUtils.renderIntoDocument(
          <Tabs onTabChange={ clickSpy }>
            <Tab title='Tab Title 1' tabId='uniqueid1'>
              <Textbox name='foo' />
              <Textbox name='bar' />
            </Tab>
          </Tabs>
        );

        instance._window = {
          history: {
            replaceState: () => {}
          },
          location: ''
        };

        instance.handleTabClick({ target: { dataset: { tabid: 'foo' } } });
        expect(clickSpy).toHaveBeenCalledWith('foo');
      });
    });
  });

  describe('mainClasses', () => {
    describe('when passing custom classNames', () => {
      it('adds it to the classList', () => {
        instance = TestUtils.renderIntoDocument(
          <Tabs className='1tab' initialTabId='uniqueid2'>
            <Tab title='Tab Title 1' tabId='uniqueid1'>
              <Textbox name='foo' />
              <Textbox name='bar' />
            </Tab>
          </Tabs>
        );

        expect(instance.mainClasses).toEqual('carbon-tabs carbon-tabs__position-top 1tab');
      });
    });

    describe('when passing a position prop', () => {
      it('adds it to the classList', () => {
        instance = TestUtils.renderIntoDocument(
          <Tabs position='left'>
            <Tab title='Tab Title 1' tabId='uniqueid1'>
              <Textbox name='foo' />
              <Textbox name='bar' />
            </Tab>
          </Tabs>
        );

        expect(instance.mainClasses).toEqual('carbon-tabs carbon-tabs__position-left');
      });
    });
  });

  describe('tabHeaderClasses', () => {
    it('adds the className included in the props to the tab', () => {
      const secondTab = TestUtils.scryRenderedDOMComponentsWithTag(instance, 'li')[1];
      expect(secondTab.classList.contains('headerClass2')).toBeTruthy();
    });

    it('adds the className included in the props as an array to the tab', () => {
      const secondTab = TestUtils.scryRenderedDOMComponentsWithTag(instance, 'li')[2];
      expect(secondTab.classList.contains('headerClass3')).toBeTruthy();
      expect(secondTab.classList.contains('headerClass4')).toBeTruthy();
    });

    describe('when tab is inValid', () => {
      it('adds a error class to the header', () => {
        instance.setState({ tabValidity: Immutable.fromJS({ uniqueid2: false }) });
        const secondTab = TestUtils.scryRenderedDOMComponentsWithTag(instance, 'li')[1];
        expect(secondTab.className).toEqual(
          'carbon-tabs__headers__header headerClass2 carbon-tabs__headers__header--error'
        );
      });
    });

    describe('when tab has a warning', () => {
      it('adds a warning class to the header', () => {
        instance.setState({ tabWarning: Immutable.fromJS({ uniqueid2: true }) });
        const secondTab = TestUtils.scryRenderedDOMComponentsWithTag(instance, 'li')[1];
        expect(secondTab.className).toEqual(
          'carbon-tabs__headers__header headerClass2 carbon-tabs__headers__header--warning'
        );
      });

      describe('when tab has an error as well', () => {
        it('does not add a warning class', () => {
          instance.setState({
            tabWarning: Immutable.fromJS({ uniqueid2: true }),
            tabValidity: Immutable.fromJS({ uniqueid2: false })
          });
          const secondTab = TestUtils.scryRenderedDOMComponentsWithTag(instance, 'li')[1];
          expect(secondTab.classList.contains('carbon-tabs__headers__header--warning')).toBeFalsy();
        });
      });
    });
  });

  describe('tabHeaders', () => {
    it('adds a data-tabid to each list item', () => {
      expect(instance.tabHeaders.props.children[0].props['data-tabid']).toEqual('uniqueid1');
    });
  });

  describe('visibleTab', () => {
    beforeEach(() => {
      instance = TestUtils.renderIntoDocument(
        <Tabs renderHiddenTabs={ false }>
          <Tab title='Tab Title 1' tabId='uniqueid1'>
            <Textbox name='foo' />
            <Textbox name='bar' />
          </Tab>
          <Tab title='Tab Title 2' tabId='uniqueid2'>
            <Textbox name='baz' />
            <Textbox name='bax' />
          </Tab>
        </Tabs>
      );
    });

    it('returns the currently visible tab', () => {
      const tab = instance.visibleTab;
      expect(tab.props.title).toEqual('Tab Title 1');
      expect(tab.props.tabId).toEqual('uniqueid1');
    });

    it('sets isTabSelected to true', () => {
      expect(instance.visibleTab.props.isTabSelected).toEqual(true);
    });
  });

  describe('tabs', () => {
    describe('when renderHiddenTabs is set to true', () => {
      it('returns an array of all child components', () => {
        expect(instance.tabs.length).toEqual(3);
      });

      it('adds a selected class to the visible tab', () => {
        expect(instance.tabs[0].props.isTabSelected).toEqual(true);
      });

      it('adds a hidden class to all other tabs', () => {
        expect(instance.tabs[1].props.isTabSelected).toEqual(false);
      });

      describe('when passed a null child', () => {
        it('ignores the null child', () => {
          const tabs = TestUtils.scryRenderedDOMComponentsWithClass(instanceWithNull, 'carbon-tab');
          expect(tabs.length).toEqual(1);
        });
      });
    });

    describe('when renderHiddenTabs is set to false', () => {
      it('returns a single child tab component', () => {
        const wrapper = render({ renderHiddenTabs: false });
        expect(wrapper.children()).toHaveLength(1);
      });
    });

    describe('when there is one child', () => {
      it('renders the only child', () => {
          const tabs = TestUtils.scryRenderedDOMComponentsWithClass(instanceOneChild, 'carbon-tab');
          expect(tabs.length).toEqual(1);
      });
    });
  });

  describe('when rendered', () => {
    it('renders its children and tabsHeader', () => {
      const wrapper = render();
      expect(wrapper.children()).toHaveLength(4);
      expect(wrapper.children().find(Tab)).toHaveLength(3);
      expect(wrapper.children().find(TabsHeader)).toHaveLength(1);
    });
  });

  describe("handleTabClick() tab trigger isn't going to happen on any keypress", () => {
    it("doesn't trigger", () => {
      const ev = {
        which: 1,
        type: 'keydown',
        target: {
          dataset: {}
        }
      };
      spyOn(instance, 'updateVisibleTab');
      instance.handleTabClick(ev);
      expect(instance.updateVisibleTab).not.toHaveBeenCalled();
    });
  });

  describe('tags', () => {
    describe('on component', () => {
      const wrapper = shallow(
        <Tabs data-element='bar' data-role='baz'>
          <Tab tabId='1' title='Test' />
        </Tabs>
      );

      it('include correct component, element and role data tags', () => {
        rootTagTest(wrapper, 'tabs', 'bar', 'baz');
      });
    });
  });

  describe('Keyboard navigation', () => {
    let wrapper, replaceSpy;

    describe('when the orientation is horizontal', () => {
      beforeEach(() => {
        wrapper = mount(
          <Tabs>
            <Tab tabId='tab1' title='Test 1' />
            <Tab tabId='tab2' title='Test 2' />
            <Tab tabId='tab3' title='Test 3' />
          </Tabs>
        );
        replaceSpy = jest.fn();
        wrapper.instance()._window = {
          history: {
            replaceState: replaceSpy
          },
          location: {
            origin: 'foo',
            pathname: 'bar'
          }
        };
      });

      describe('when pressing the right arrow', () => {
        it('focuses on the next right tab and loops back round to the first tab', () => {
          wrapper.setState({ selectedTabId: 'tab1' });
          wrapper
            .find('.carbon-tabs__headers__header--selected')
            .instance()
            .focus();
          expect(Browser.getActiveElement().getAttribute('data-tabid')).toEqual('tab1');
          wrapper
            .find('.carbon-tabs__headers__header--selected')
            .simulate('keyDown', { key: 'ArrowRight', which: 39, stopPropagation: () => {} });
          expect(Browser.getActiveElement().getAttribute('data-tabid')).toEqual('tab2');
          expect(replaceSpy).toHaveBeenCalledWith(null, 'change-tab', 'foobar#tab2');
          wrapper
            .find('.carbon-tabs__headers__header--selected')
            .simulate('keyDown', { key: 'ArrowRight', which: 39, stopPropagation: () => {} });
          expect(Browser.getActiveElement().getAttribute('data-tabid')).toEqual('tab3');
          expect(replaceSpy).toHaveBeenCalledWith(null, 'change-tab', 'foobar#tab3');
          wrapper
            .find('.carbon-tabs__headers__header--selected')
            .simulate('keyDown', { key: 'ArrowRight', which: 39, stopPropagation: () => {} });
          expect(replaceSpy.mock.calls.length).toEqual(3);
          expect(Browser.getActiveElement().getAttribute('data-tabid')).toEqual('tab1');
        });
      });

      describe('when pressing the left arrow', () => {
        it('focuses on the next left tab and loops back round to the last tab', () => {
          wrapper.setState({ selectedTabId: 'tab3' });
          wrapper
            .find('.carbon-tabs__headers__header--selected')
            .instance()
            .focus();
          expect(Browser.getActiveElement().getAttribute('data-tabid')).toEqual('tab3');
          wrapper
            .find('.carbon-tabs__headers__header--selected')
            .simulate('keyDown', { key: 'ArrowLeft', which: 37, stopPropagation: () => {} });
          expect(replaceSpy).toHaveBeenCalledWith(null, 'change-tab', 'foobar#tab2');
          expect(Browser.getActiveElement().getAttribute('data-tabid')).toEqual('tab2');
          wrapper
            .find('.carbon-tabs__headers__header--selected')
            .simulate('keyDown', { key: 'ArrowLeft', which: 37, stopPropagation: () => {} });
          expect(replaceSpy).toHaveBeenCalledWith(null, 'change-tab', 'foobar#tab1');
          wrapper
            .find('.carbon-tabs__headers__header--selected')
            .simulate('keyDown', { key: 'ArrowLeft', which: 37, stopPropagation: () => {} });
          expect(replaceSpy.mock.calls.length).toEqual(3);
          expect(Browser.getActiveElement().getAttribute('data-tabid')).toEqual('tab3');
        });
      });

      describe('when pressing the up key', () => {
        it('doesnt do anything', () => {
          wrapper
            .find('.carbon-tabs__headers__header--selected')
            .simulate('keyDown', { key: 'UpArrow', which: 38, stopPropagation: () => {} });
          expect(replaceSpy).not.toHaveBeenCalled();
        });
      });

      describe('when pressing the down key', () => {
        it('doesnt do anything', () => {
          wrapper
            .find('.carbon-tabs__headers__header--selected')
            .simulate('keyDown', { key: 'DownArrow', which: 40, stopPropagation: () => {} });
          expect(replaceSpy).not.toHaveBeenCalled();
        });
      });
    });

    describe('when the orientation is vertical', () => {
      beforeEach(() => {
        wrapper = mount(
          <Tabs position='left'>
            <Tab tabId='tab1' title='Test 1' />
            <Tab tabId='tab2' title='Test 2' />
            <Tab tabId='tab3' title='Test 3' />
          </Tabs>
        );
        replaceSpy = jest.fn();
        wrapper.instance()._window = {
          history: {
            replaceState: replaceSpy
          },
          location: {
            origin: 'foo',
            pathname: 'bar'
          }
        };
      });

      describe('when pressing the down arrow', () => {
        it('focuses on the next below tab and loops back round to the top tab', () => {
          wrapper.setState({ selectedTabId: 'tab1' });
          wrapper
            .find('.carbon-tabs__headers__header--selected')
            .instance()
            .focus();
          expect(Browser.getActiveElement().getAttribute('data-tabid')).toEqual('tab1');
          wrapper
            .find('.carbon-tabs__headers__header--selected')
            .simulate('keyDown', { key: 'ArrowDown', which: 40, stopPropagation: () => {} });
          expect(Browser.getActiveElement().getAttribute('data-tabid')).toEqual('tab2');
          expect(replaceSpy).toHaveBeenCalledWith(null, 'change-tab', 'foobar#tab2');
          wrapper
            .find('.carbon-tabs__headers__header--selected')
            .simulate('keyDown', { key: 'ArrowDown', which: 40, stopPropagation: () => {} });
          expect(Browser.getActiveElement().getAttribute('data-tabid')).toEqual('tab3');
          expect(replaceSpy).toHaveBeenCalledWith(null, 'change-tab', 'foobar#tab3');
          wrapper
            .find('.carbon-tabs__headers__header--selected')
            .simulate('keyDown', { key: 'ArrowDown', which: 40, stopPropagation: () => {} });
          expect(replaceSpy.mock.calls.length).toEqual(3);
          expect(Browser.getActiveElement().getAttribute('data-tabid')).toEqual('tab1');
        });
      });

      describe('when pressing the up arrow', () => {
        it('focuses on the next above tab and loops back round to the bottom tab', () => {
          wrapper.setState({ selectedTabId: 'tab3' });
          wrapper
            .find('.carbon-tabs__headers__header--selected')
            .instance()
            .focus();
          expect(Browser.getActiveElement().getAttribute('data-tabid')).toEqual('tab3');
          wrapper
            .find('.carbon-tabs__headers__header--selected')
            .simulate('keyDown', { key: 'ArrowUp', which: 38, stopPropagation: () => {} });
          expect(replaceSpy).toHaveBeenCalledWith(null, 'change-tab', 'foobar#tab2');
          expect(Browser.getActiveElement().getAttribute('data-tabid')).toEqual('tab2');
          wrapper
            .find('.carbon-tabs__headers__header--selected')
            .simulate('keyDown', { key: 'ArrowUp', which: 38, stopPropagation: () => {} });
          expect(replaceSpy).toHaveBeenCalledWith(null, 'change-tab', 'foobar#tab1');
          wrapper
            .find('.carbon-tabs__headers__header--selected')
            .simulate('keyDown', { key: 'ArrowUp', which: 38, stopPropagation: () => {} });
          expect(replaceSpy.mock.calls.length).toEqual(3);
          expect(Browser.getActiveElement().getAttribute('data-tabid')).toEqual('tab3');
        });
      });

      describe('when pressing the left key', () => {
        it('doesnt do anything', () => {
          wrapper
            .find('.carbon-tabs__headers__header--selected')
            .simulate('keyDown', { key: 'LeftArrow', which: 37, stopPropagation: () => {} });
          expect(replaceSpy).not.toHaveBeenCalled();
        });
      });

      describe('when pressing the right key', () => {
        it('doesnt do anything', () => {
          wrapper
            .find('.carbon-tabs__headers__header--selected')
            .simulate('keyDown', { key: 'RightArrow', which: 39, stopPropagation: () => {} });
          expect(replaceSpy).not.toHaveBeenCalled();
        });
      });
    });
  });
});
