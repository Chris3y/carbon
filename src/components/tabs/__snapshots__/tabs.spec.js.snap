// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs when rendered matches the snaphot 1`] = `
.c3 {
  display: none;
  display: block;
}

.c0 {
  margin-top: 15px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-shadow: inset 0px -2px 0px 0px #E5EAEC;
  cursor: pointer;
  list-style: none;
  margin: 0 0 10px;
  padding: 0;
}

.c2 {
  background-color: transparent;
  border-bottom: 2px solid #E5EAEC;
  color: rgba(0,0,0,0.55);
  display: inline-block;
  font-weight: bold;
  height: 100%;
  margin-left: 2px;
  padding: 11px 15px 10px;
  color: rgba(0,0,0,0.9);
  background-color: transparent;
  border-bottom-color: #26A826;
}

.c2:first-child {
  margin-left: 0;
}

.c2:focus,
.c2:hover {
  background: transparent;
  border-bottom-color: #006300;
  color: rgba(0,0,0,0.9);
  outline: none;
}

.c2:focus {
  outline: none;
}

.c2:hover {
  background: transparent;
  border-bottom-color: #26A826;
  color: rgba(0,0,0,0.9);
}

<div
  className="c0"
  data-component="tabs"
>
  <ul
    className="c1"
    role="tablist"
  >
    <li
      aria-selected={true}
      className="c2"
      data-component="tab-header"
      data-element="select-tab"
      data-tabid="uniqueid1"
      id="uniqueid1-tab"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="tab"
      tabIndex="0"
    >
      Tab Title 1
    </li>
  </ul>
  <div
    aria-labelledby="uniqueid1-tab"
    className="c3"
    data-component="tab"
    role="tabPanel"
  >
    Tab Content
  </div>
</div>
`;
